Sources/Commands/Utilities/SymbolGraphExtract.swift

```diff
@@ -48,12 +57,115 @@ public struct SymbolGraphExtract {
         outputRedirection: TSCBasic.Process.OutputRedirection = .none,
         outputDirectory: AbsolutePath,
         verboseOutput: Bool
+    ) throws {
+        if target.underlyingTarget is SwiftTarget {
+            return try extractSymbolGraphFromSwiftTarget(
+                target: target,
+                buildPlan: buildPlan,
+                outputRedirection: outputRedirection,
+                outputDirectory: outputDirectory.appending(component: "swift"),
+                verboseOutput: verboseOutput
+            )
+        } else if let clangTarget = target.underlyingTarget as? ClangTarget,
+                  case let .clang(buildDescription) = buildPlan.targetMap[target]
+        {
+            return try extractSymbolGraphFromClangTarget(
+                target: clangTarget,
+                buildDescription: buildDescription,
+                outputRedirection: outputRedirection,
+                outputDirectory: outputDirectory.appending(component: "clang"),
+                verboseOutput: verboseOutput
+            )
+        }
+    }
+    
+    func extractSymbolGraphFromClangTarget(
+        target: ClangTarget,
+        buildDescription: ClangTargetBuildDescription,
+        outputRedirection: Process.OutputRedirection,
+        outputDirectory: AbsolutePath,
+        verboseOutput: Bool
+    ) throws {
+        // Collect the set of public headers by filtering all of the target's headers to those
+        // inside the target's include directory.
+        var publicHeaders = Set(target.headers.filter(target.includeDir.isAncestor))
+        
+        guard !publicHeaders.isEmpty else {
+            return
+        }
+        
+        let umbrellaHeader: AbsolutePath?
+        switch target.moduleMapType {
+        case .umbrellaHeader(let umbrellaHeaderFile):
+            // The target is configured with an umbrella header – store its path separately
+            // and remove it from the general list of public headers.
+            umbrellaHeader = umbrellaHeaderFile
+            publicHeaders.remove(umbrellaHeaderFile)
+        case .umbrellaDirectory(_):
+            // TODO: Support symbol graph extraction for umbrella directory module map type
+            fallthrough
+        case .custom(_):
+            // TODO: Support symbol graph extraction for custom module map file type
+            fallthrough
```

Comment:
I don't think anything special needs to be done for umbrella directory layouts, since ExtractAPI works very similar to module builds. For a custom module map that is a bit more complicated, what information do we have about the custom module map?

Reply:
> For a custom module map that is a bit more complicated, what information do we have about the custom module map?

I think we just have the path to a `module.modulemap` file. This _might_ be excuslively for system libraries actually: https://github.com/apple/swift-package-manager/blob/263171977ebcd47f4aaca1202cff5a96c5158a64/Documentation/Usage.md#import-system-libraries.

Reply:
I think this is only for system libraries as well, since this doesn't actually build anything if I read the docs right. I would just ignore it, documenting  C/ObjC system library dependencies is outside the scope of this effort.